File: docker/tests/validate_rest.py
import json
import sys
import time
import requests

BASE_URL = "http://localhost:3000"
API_BASE = f"{BASE_URL}/signalk/v1/api"
AUTH_BASE = f"{BASE_URL}/signalk/v1/auth"
WS_URL = "ws://localhost:3000/signalk/v1/stream"

ADMIN_USER = "admin"
ADMIN_PASS = "test123"

TEST_POSITION = {"latitude": 48.1173, "longitude": 11.5167}


def create_admin():
    """Create admin account via enableSecurity endpoint."""
    r = requests.post(
        f"{API_BASE}/enableSecurity",
        json={"userId": ADMIN_USER, "password": ADMIN_PASS},
        timeout=5,
    )
    if r.status_code in (200, 201, 202):
        return True
    print(f"  enableSecurity HTTP {r.status_code}: {r.text[:200]}")
    return False


def login():
    """Login and return JWT token."""
    r = requests.post(
        f"{AUTH_BASE}/login",
        json={"username": ADMIN_USER, "password": ADMIN_PASS},
        timeout=5,
    )
    if r.status_code == 200:
        return r.json().get("token")
    print(f"  login HTTP {r.status_code}: {r.text[:200]}")
    return None


def send_position_delta(token):
    """Send test position data via WebSocket delta."""
    from websocket import create_connection

    ws = create_connection(f"{WS_URL}?subscribe=none&token={token}", timeout=10)
    hello = json.loads(ws.recv())
    print(f"  Connected (server: {hello.get('name', 'unknown')})")

    delta = {
        "context": "vessels.self",
        "updates": [
            {
                "source": {"label": "pre-release-test"},
                "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                "values": [
                    {"path": "navigation.position", "value": TEST_POSITION}
                ],
            }
        ],
    }
    ws.send(json.dumps(delta))
    time.sleep(1)
    ws.close()
    return True


def validate_position(timeout=10):
    """Check position data via REST API."""
    url = f"{API_BASE}/vessels/self/navigation/position/value"
    start = time.time()
    while time.time() - start < timeout:
        try:
            r = requests.get(url, timeout=2)
            if r.status_code == 200:
                data = r.json()
                if data.get("latitude") is not None and data.get("longitude") is not None:
                    return data
        except Exception:
            pass
        time.sleep(0.5)
    return None


print("SignalK Pre-Release REST Validation")
print("=" * 40)

print("\n1. Creating admin account...")
if not create_admin():
    print("FAIL: Could not create admin account")
    sys.exit(1)
print("   OK")

print("\n2. Logging in...")
token = login()
if not token:
    print("FAIL: Could not login")
    sys.exit(1)
print("   OK")

print("\n3. Sending position delta via WebSocket...")
if not send_position_delta(token):
    print("FAIL: Could not send delta")
    sys.exit(1)
print("   OK")

print("\n4. Validating position via REST API...")
value = validate_position()
if not value:
    print("FAIL: Position not available in REST API")
    try:
        r = requests.get(f"{API_BASE}/vessels/self", timeout=5)
        print(f"  /vessels/self -> HTTP {r.status_code}")
        print(f"  {r.text[:500]}")
    except Exception as e:
        print(f"  /vessels/self -> {e}")
    sys.exit(1)

lat = value["latitude"]
lon = value["longitude"]
print(f"   Position: lat={lat}, lon={lon}")

if abs(lat - TEST_POSITION["latitude"]) > 0.001 or abs(lon - TEST_POSITION["longitude"]) > 0.001:
    print(f"FAIL: Position mismatch (expected {TEST_POSITION})")
    sys.exit(1)

print("\nREST validation PASSED")

File: docker/tests/validate_rest.js
const WebSocket = require('ws');

const BASE_URL = 'http://localhost:3000';
const SK_SERVER = `${BASE_URL}/skServer`;
const API_BASE = `${BASE_URL}/signalk/v1/api`;
const AUTH_BASE = `${BASE_URL}/signalk/v1/auth`;
const WS_URL = 'ws://localhost:3000/signalk/v1/stream';

const ADMIN_USER = 'admin';
const ADMIN_PASS = 'test123';
const TEST_POSITION = { latitude: 48.1173, longitude: 11.5167 };

async function createAdmin() {
  const r = await fetch(`${SK_SERVER}/enableSecurity`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ userId: ADMIN_USER, password: ADMIN_PASS, type: 'admin' }),
  });
  if (r.ok) return true;
  console.log(`  enableSecurity HTTP ${r.status}: ${(await r.text()).slice(0, 200)}`);
  return false;
}

async function login() {
  const r = await fetch(`${AUTH_BASE}/login`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ username: ADMIN_USER, password: ADMIN_PASS }),
  });
  if (r.ok) {
    const data = await r.json();
    return data.token;
  }
  console.log(`  login HTTP ${r.status}: ${(await r.text()).slice(0, 200)}`);
  return null;
}

function sendPositionDelta(token) {
  return new Promise((resolve, reject) => {
    const ws = new WebSocket(`${WS_URL}?subscribe=none&token=${token}`);
    const timer = setTimeout(() => { ws.close(); reject(new Error('WebSocket timeout')); }, 10000);

    ws.on('message', (raw) => {
      const hello = JSON.parse(raw.toString());
      console.log(`  Connected (server: ${hello.name || 'unknown'})`);

      const delta = {
        context: 'vessels.self',
        updates: [{
          source: { label: 'pre-release-test' },
          timestamp: new Date().toISOString(),
          values: [{ path: 'navigation.position', value: TEST_POSITION }],
        }],
      };
      ws.send(JSON.stringify(delta));
      setTimeout(() => { clearTimeout(timer); ws.close(); resolve(true); }, 1000);
    });

    ws.on('error', (err) => { clearTimeout(timer); reject(err); });
  });
}

async function validatePosition(timeout = 10000) {
  const url = `${API_BASE}/vessels/self/navigation/position/value`;
  const start = Date.now();
  while (Date.now() - start < timeout) {
    try {
      const r = await fetch(url, { signal: AbortSignal.timeout(2000) });
      if (r.ok) {
        const data = await r.json();
        if (data.latitude != null && data.longitude != null) return data;
      }
    } catch {}
    await new Promise((r) => setTimeout(r, 500));
  }
  return null;
}

async function main() {
  console.log('SignalK Pre-Release REST Validation');
  console.log('='.repeat(40));

  console.log('\n1. Creating admin account...');
  if (!(await createAdmin())) {
    console.log('FAIL: Could not create admin account');
    process.exit(1);
  }
  console.log('   OK');

  console.log('\n2. Logging in...');
  const token = await login();
  if (!token) {
    console.log('FAIL: Could not login');
    process.exit(1);
  }
  console.log('   OK');

  console.log('\n3. Sending position delta via WebSocket...');
  try {
    await sendPositionDelta(token);
  } catch (err) {
    console.log(`FAIL: Could not send delta: ${err.message}`);
    process.exit(1);
  }
  console.log('   OK');

  console.log('\n4. Validating position via REST API...');
  const value = await validatePosition();
  if (!value) {
    console.log('FAIL: Position not available in REST API');
    try {
      const r = await fetch(`${API_BASE}/vessels/self`, { signal: AbortSignal.timeout(5000) });
      console.log(`  /vessels/self -> HTTP ${r.status}`);
      console.log(`  ${(await r.text()).slice(0, 500)}`);
    } catch (e) {
      console.log(`  /vessels/self -> ${e.message}`);
    }
    process.exit(1);
  }

  const { latitude: lat, longitude: lon } = value;
  console.log(`   Position: lat=${lat}, lon=${lon}`);

  if (Math.abs(lat - TEST_POSITION.latitude) > 0.001 || Math.abs(lon - TEST_POSITION.longitude) > 0.001) {
    console.log(`FAIL: Position mismatch (expected ${JSON.stringify(TEST_POSITION)})`);
    process.exit(1);
  }

  console.log('\nREST validation PASSED');
}

main().catch((err) => {
  console.error('Unexpected error:', err);
  process.exit(1);
});
